src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,987 @@ mod v1 {
+            if !self.exists() {
+                return Err(box_err!("snapshot file {} not exist", self.path()));
+            }
+            for cf_file in &mut self.cf_files {
+                // initialize cf file size and reader
+                let size = try!(get_file_size(&cf_file.path));
+                let file = try!(File::open(&cf_file.path));
+                cf_file.size = size;
+                cf_file.file = Some(file);
+            }
+            self.load_snapshot_metadata()
+        }
+
+        fn init_for_receiving(&mut self) -> io::Result<()> {
+            if self.exists() || self.meta_file.file.is_some() {
+                return Ok(());
+            }
+
+            for cf_file in &mut self.cf_files {
+                let f =
+                    try!(OpenOptions::new().write(true).create_new(true).open(&cf_file.tmp_path));
+                cf_file.file = Some(f);
+            }
+            let f = try!(OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(&self.meta_file.tmp_path));
+            self.meta_file.file = Some(f);
+            Ok(())
+        }
+
+        fn init_for_applying(&mut self) -> RaftStoreResult<()> {
+            self.load_snapshot_metadata()
+        }
+
+        fn load_snapshot_metadata(&mut self) -> RaftStoreResult<()> {
+            let size = try!(get_file_size(&self.meta_file.path));
+            let mut file = try!(File::open(&self.meta_file.path));
+            let mut buf = Vec::with_capacity(size as usize);
+            try!(file.read_to_end(&mut buf));
+            let mut snapshot_data = RaftSnapshotData::new();
+            try!(snapshot_data.merge_from_bytes(&buf));
+            self.meta_file.data = snapshot_data;
+            Ok(())
+        }
+
+        fn get_display_path(dir_path: &PathBuf, prefix: &str) -> String {
+            let mut cf_names = String::from("");
+            for (i, cf) in get_snapshot_cfs().iter().enumerate() {
+                if i == 0 {
+                    cf_names += "(";
+                    cf_names += cf;
+                } else {
+                    cf_names += "|";
+                    cf_names += cf;
+                }
+            }
+            cf_names += ")";
+            format!("{}/{}_{}{}",
+                    dir_path.as_path().to_str().unwrap().to_owned(),
+                    prefix,
+                    cf_names,
+                    SST_FILE_SUFFIX)
+        }
+
+        fn try_delete(&self) -> io::Result<()> {
+            debug!("deleting {}", self.path());
+            let exists = self.exists();
+            let mut total_size = 0;
+            for cf_file in &self.cf_files {
+                if file_exists(&cf_file.tmp_path) {
+                    delete_file(&cf_file.tmp_path);
+                }
+                if exists {
+                    total_size += try!(get_file_size(&cf_file.path));
+                }
+                delete_file(&cf_file.path);
```

Comment:
try_delete need called when new_for_building or new_for_sending failed. Further more, try_delete need check if sst_writer is closed normally.